src/haz3lcore/statics/Statics.re

```diff
@@ -47,13 +47,26 @@ let map_m = (f, xs, m: Map.t) =>
 let add_info = (ids: list(Id.t), info: Info.t, m: Map.t): Map.t =>
   ids |> List.fold_left((m, id) => Id.Map.add(id, info, m), m);
 
-let extend_let_def_ctx =
-    (ctx: Ctx.t, pat: UPat.t, pat_ctx: Ctx.t, def: UExp.t): Ctx.t =>
-  if (UPat.is_tuple_of_arrows(pat) && UExp.is_tuple_of_functions(def)) {
-    pat_ctx;
-  } else {
-    ctx;
+let is_recursive = (ctx, p, def, syn: Typ.t) => {
+  switch (Term.UPat.get_num_of_vars(p), Term.UExp.get_num_of_functions(def)) {
+  | (Some(num_vars), Some(num_fns))
+      when num_vars != 0 && num_vars == num_fns =>
+    switch (Typ.normalize(ctx, syn)) {
```

Comment:
The process of determining recursiveness doesn't consider whether the type of the definition (the function literal) matches the annotation. I believe this has already been addressed properly...?
![image](https://github.com/hazelgrove/hazel/assets/108375845/abc4c8c2-3366-4f1d-8ba5-3fa4010dd84b)